library/src/updater.rs

```diff
@@ -51,6 +50,7 @@ impl Display for UpdateStatus {
 pub enum InitError {
     InvalidArgument(String, String),
     AlreadyInitialized,
+    FailedToCleanUpFailedPatch,
```

Comment:
I'm not sure the contract here.  The first one is fatal (you can't continue using updater until you call init correctly), the second two are benign, right?  Is this differentiated to the C callers?

Reply:
Any of these being returned as an `Err` result from the `init` function results in `shorebird_init` returning false. The contract as I've been thinking about it is:

- If `shorebird_init` returns true, we can confidently say that the updater is initialized. 
- If it returns false, that particular call to init was not successful, but that doesn't necessarily mean that the updater is not ready to use (e.g., in the case of `AlreadyInitialized`). I would view the failure to clean up a failed patch as less benign than `AlreadyInitialized`, as it could mean that the bad patch is still in place as our next boot patch.

Not sure if that explanation helps. Maybe there's a better name for this or missing docs?

Reply:
I'm just wondering about the C++ code:
https://github.com/shorebirdtech/engine/blob/shorebird/dev/shell/common/shorebird/shorebird.cc#L128

Right now it looks like it doesn't even check the return value.  I'm just not sure what behavior it should have if it did.
Should it continue to call shorebird_ methods after a false return?  Does it need to check some other value in that case?  e.g. consider the case of the FCM multi-threaded call.  When the second thread calls shorebird_init and gets back false (currently ignored).  It presumably should keep calling shorebird methods?  But if any thread ever got the "failed to cleanup failed patch" response, is it allowed to keep calling shorebird methods?  Certainly in the "invalid argument" case, it seems like calling more shorebird methods is folly (although I suspect that since active_path returns null, after a failed init everything is fine?)
https://github.com/shorebirdtech/engine/blob/shorebird/dev/shell/common/shorebird/shorebird.cc#L143

Reply:
Maybe the answer is that shorebird_init shouldn't even return a bool and the contract is "its always OK to keep callign shorebird methods, they'll never crash, just they might also not do anything"?

Reply:
> Maybe the answer is that shorebird_init shouldn't even return a bool and the contract is "its always OK to keep callign shorebird methods, they'll never crash, just they might also not do anything"?

I think this is the answer. We have to support this for the `shorebird_code_push` package, and it would be nice to have the same set of assumptions for interacting with the engine. Changing the return value of the `shorebird_init` C API seems like a separate change if that works for you.

Reply:
sg.  We should document this assumption then (either as part of this change or the other), that these error conditions are just advisory (they just log) and that all shorebird apis are expected to never crash/do wrong things, even after init errors.

Reply:
Done.